karpenter-operator/controllers/aws/ec2_nodeclass.go

```diff
@@ -0,0 +1,376 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"time"
+
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	hyperkarpenterv1 "github.com/openshift/hypershift/api/karpenter/v1alpha1"
+	"github.com/openshift/hypershift/karpenter-operator/controllers/karpenter/assets"
+	supportassets "github.com/openshift/hypershift/support/assets"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/upsert"
+
+	awskarpenterapis "github.com/aws/karpenter-provider-aws/pkg/apis"
+	awskarpenterv1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+
+	admissionv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/ptr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// userDataAMILabel is a label set in the userData secret generated for karpenter instances.
+	userDataAMILabel = "hypershift.openshift.io/ami"
+)
+
+var (
+	crdEC2NodeClass = supportassets.MustCRD(assets.ReadFile, "karpenter.k8s.aws_ec2nodeclasses.yaml")
+
+	crdOpenshiftEC2NodeClass = supportassets.MustCRD(assets.ReadFile, "karpenter.hypershift.openshift.io_openshiftec2nodeclasses.yaml")
+)
+
+type EC2NodeClassReconciler struct {
+	Namespace string
+
+	managementClient client.Client
+	guestClient      client.Client
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *EC2NodeClassReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, managementCluster cluster.Cluster) error {
+	r.managementClient = managementCluster.GetClient()
+	r.guestClient = mgr.GetClient()
+	r.CreateOrUpdateProvider = upsert.New(false)
+
+	// First install the CRDs so we can create a watch below.
+	if err := r.reconcileCRDs(ctx, true); err != nil {
```

Comment:
I'd expect we let the main controller handle crd installation, then this controller is just implementing mirroring

Reply:
my thoughts are to let this controller which is aws specific to install its CRDs (EC2NodeClass, etc), and then when we add support for azure, we would need a new controller which will install the azure specific CRDs (AKSNodeClass, etc).